### PR TITLE
Add the unreal2 protocol

### PR DESCRIPTION
**Pull request by Douile**
First draft of an unreal2 implementation. There's still quite a few TODOs but as it is it works with the killingfloor test I made (I have yet to test the other games that use this protocol).

This is my first protocol implementation, please let me know if I have done anything wrong.

The part that needs the most work for this to be ready is the StringDecoder implementation: unreal2 uses either latin1 or ucs2(?) so that is kind of blocked by #89 as there is already work on adding a latin1 decoder (this work in `feat/encoding-latin1` also includes a helper function for finding delimiters which would be used).

Missing from this implementation is the extra response data parsing: the server info packet can contain extra data for certain games. The node implementation detects this based on what mutators the server returns and goes back to the server info buffer and parses the extra data. I haven't looked into whether we would want to do the same thing or parse additional data based on ExtraRequestSettings like we do with valve.

Also of note is that this protocol relies heavily on timeouts, we don't know how many packets we will get in response to each query (for rules, players, and possibly server info). As a result we keep receiving packets until we get a timeout. I would like to look into whether there is a better way to do this: maybe the 4 bytes of header we ignore contain information on how many response packets to expect. If anyone has more information about this protocol please do let me know.

This was in the roadmap:
https://github.com/orgs/gamedig/projects/1?pane=issue&itemId=22463327

### CODE REVIEW

**Comment by CosminPerRam** [comment]
> This is my first protocol implementation [...]

Great work :)
I will come back with a review when I get the time to do so.

> Missing from this implementation is the extra response data parsing [...]

Not really a merging-blocking problem, all protocols that are already implemented weren't fully working/completed when I first added them (also applies to next quote response), many smaller steps are better than rare big ones, can't run without walking first.

> [...] so that is kind of blocked by https://github.com/gamedig/rust-gamedig/issues/89

Will look into it.

**Review by CosminPerRam — APPROVED** [review]

**Comment by Douile** [comment]
I noticed while implementing latin1 that some ut2003 servers send UTF8 strings when the node implementation expected latin1. When we parse as latin1 in this case the colour stripping is broken and hence random characters are left in the final name. (The same behavior as node-gamedig).

File: src/protocols/unreal2/types.rs [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+            server_id,
+            ip,
+            game_port,
+            query_port,
+            name,
+            map,
+            game_type,
+            num_players,
+            max_players,
+        })
```
Comment (CosminPerRam) [inline src/protocols/unreal2/types.rs:72]: As there isn't any usage of index skipping or multi-line processing, we could move these variable declarations directly inside the return block to make it a bit more tidier.
```rust
Ok(ServerInfo {
    server_id: buffer.read()?,
    ip: buffer.read_string::<Unreal2StringDecoder>(None)?
  //...
```

File: src/protocols/unreal2/types.rs [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+        Ok(())
+    }
+
+    /// Length of both players and bots.
+    pub fn total_len(&self) -> usize { self.players.len() + self.bots.len() }
+}
+
+/// Unreal 2 player info.
+#[derive(Clone, Debug, PartialEq)]
```
Comment (CosminPerRam) [inline src/protocols/unreal2/types.rs:163]: Make sure we have all the standard derives (`Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord`) on all types (if they can be added).

**Review by CosminPerRam — CHANGES_REQUESTED** [review]
Great job! Thanks bunch for such a big contribution.
Just a heads up to also add:
* Changes in CHANGELOG.md (Don't forget to note yourself!).
* The new games in GAMES.md (and CHANGELOG.md).
* The protocol in PROTOCOLS.md.
* Response in RESPONSES.md.

File: src/protocols/unreal2/types.rs (hunk 1 above)
Comment (Douile) [reply src/protocols/unreal2/types.rs:72]: I wasn't sure if that could cause it to execute out of order, but yeah that would be cleaner.

File: src/protocols/unreal2/types.rs (hunk 2 above)
Comment (Douile) [reply src/protocols/unreal2/types.rs]: I think I got all of them in: https://github.com/gamedig/rust-gamedig/pull/124/commits/88b3cfa27d149ea27ab63dbaa56e9ed2424db66d

File: src/protocols/unreal2/protocol.rs [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+                    return false;
+                }
+                char_skip = char_skip.saturating_sub(1);
+
+                char_skip == 0
+            })
+            .collect();
+
+        // Remove all characters between 0x00 and 0x1a
+        let result = result.replace(|c: char| c > '\x00' && c <= '\x1a', "");
```
Comment (Douile) [inline src/protocols/unreal2/protocol.rs:288]: I'm not sure this belongs in the string decoder, do you think I should move it to a separate helper?

File: RESPONSES.tgn
Comment (Douile) [inline RESPONSES.tgn:1]: Let me know whether I should remove this file.

Comment (CosminPerRam) [reply RESPONSES.tgn:1]: Please do so.

File: src/protocols/unreal2/protocol.rs (hunk 3 above)
Comment (CosminPerRam) [reply src/protocols/unreal2/protocol.rs:288]: As I think this file would get refactored/modified multiple times, we could leave it here for now.

**Pull request merged**